internal/storage/sbom_store.go

```diff
@@ -0,0 +1,525 @@
+	}
+
+	options := &generic.StoreOptions{RESTOptions: optsGetter, AttrFunc: GetAttrs}
+	if err := store.CompleteWithOptions(options); err != nil {
+		return nil, fmt.Errorf("unable to complete store with options: %w", err)
+	}
+
+	return store, nil
+}
+
+var _ storage.Interface = &sbomStore{}
+
+type sbomStore struct {
+	broadcaster *watch.Broadcaster
+	db          *sqlx.DB
+}
+
+// Returns Versioner associated with this interface.
+func (s *sbomStore) Versioner() storage.Versioner {
+	return storage.APIObjectVersioner{}
+}
+
+// Create adds a new object at a key unless it already exists. 'ttl' is time-to-live
+// in seconds (0 means forever). If no error is returned and out is not nil, out will be
+// set to the read value from database.
+func (s *sbomStore) Create(ctx context.Context, key string, obj, out runtime.Object, _ uint64) error {
+	name, namespace := extractNameAndNamespace(key)
+	if name == "" || namespace == "" {
+		return storage.NewInternalErrorf("invalid key: %s", key)
+	}
+
+	sbom, ok := obj.(*v1alpha1.SBOM)
+	if !ok {
+		return storage.NewInvalidObjError(key, fmt.Sprintf("unexpected object type: %T", obj))
+	}
+
+	if err := s.Versioner().UpdateObject(obj, 1); err != nil {
+		return storage.NewInternalError(err.Error())
+	}
+
+	bytes, err := json.Marshal(sbom)
+	if err != nil {
+		return storage.NewInternalError(err.Error())
+	}
+
+	query, args, err := sq.Insert("sboms").
+		Columns("name", "namespace", "object").
+		Values(name, namespace, bytes).
+		ToSql()
+	if err != nil {
+		return storage.NewInternalError(err.Error())
+	}
+
+	_, err = s.db.ExecContext(ctx, query, args...)
+	if err != nil {
+		return storage.NewInternalError(err.Error())
+	}
+
+	outSBOM, ok := out.(*v1alpha1.SBOM)
+	if !ok {
+		return storage.NewInvalidObjError(key, fmt.Sprintf("unexpected out object type: %T", out))
+	}
+
+	*outSBOM = *sbom
+
+	if err := s.broadcaster.Action(watch.Added, sbom); err != nil {
+		return storage.NewInternalError(err.Error())
+	}
+
+	return nil
+}
+
+// Delete removes the specified key and returns the value that existed at that spot.
+// If key didn't exist, it will return NotFound storage error.
+// If 'cachedExistingObject' is non-nil, it can be used as a suggestion about the
+// current version of the object to avoid read operation from storage to get it.
+// However, the implementations have to retry in case suggestion is stale.
+func (s *sbomStore) Delete(
+	ctx context.Context, key string, out runtime.Object, preconditions *storage.Preconditions,
+	validateDeletion storage.ValidateObjectFunc, _ runtime.Object,
+) error {
+	name, namespace := extractNameAndNamespace(key)
+	if name == "" || namespace == "" {
+		return storage.NewInternalErrorf("invalid key: %s", key)
+	}
+
+	tx, err := s.db.BeginTxx(ctx, nil)
+	if err != nil {
+		return storage.NewInternalError(err.Error())
+	}
+	defer func() {
+		if err := tx.Rollback(); !errors.Is(err, sql.ErrTxDone) {
```

Comment:
So we always try to roolback, but if the transaction was successfully committed we get `sql.ErrTxDone` and we just ignore that, because it's all good. Do I get it right?

Reply:
yes this is a common pattern, see: https://go.dev/doc/database/execute-transactions

The difference is that we check the error returned